x/group/keeper/keeper.go

```diff
@@ -272,30 +264,87 @@ func (k Keeper) iterateProposalsByVPEnd(ctx sdk.Context, cb func(proposal group.
 	return nil
 }
 
-func (k Keeper) UpdateTallyOfVPEndProposals(ctx sdk.Context) error {
-	k.iterateProposalsByVPEnd(ctx, func(proposal group.Proposal) (bool, error) {
+// pruneProposal deletes a proposal from state.
+func (k Keeper) pruneProposal(ctx sdk.Context, proposalID uint64) error {
+	store := ctx.KVStore(k.key)
 
-		policyInfo, err := k.getGroupPolicyInfo(ctx, proposal.Address)
+	err := k.proposalTable.Delete(store, proposalID)
+	if err != nil {
+		return err
+	}
+
+	k.Logger(ctx).Debug(fmt.Sprintf("Pruned proposal %d", proposalID))
+	return nil
+}
+
+// pruneVotes prunes all votes for a proposal from state.
+func (k Keeper) pruneVotes(ctx sdk.Context, proposalID uint64) error {
+	store := ctx.KVStore(k.key)
+	it, err := k.voteByProposalIndex.Get(store, proposalID)
+	if err != nil {
+		return err
+	}
+	defer it.Close()
+
+	var vote group.Vote
```

Comment:
should we move this into the loop too?
same thing in `validateDecisionPolicies`

Reply:
Good catch! One day we'll also need to audit through all other modules...

Also, it seems to me that it's only a bug on array fields. So proto messages that don't have array fields are safe. but I'm not 100% sure.

Reply:
I see, but indeed if we're not 100% sure, best to use the same pattern that we're sure doesn't cause issues everywhere